datafusion/expr/src/built_in_function.rs

```diff
@@ -431,29 +452,21 @@ impl BuiltinScalarFunction {
 
 impl fmt::Display for BuiltinScalarFunction {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        for (func_name, func) in NAME_TO_FUNCTION.iter() {
-            if func == self {
-                return write!(f, "{}", func_name);
-            }
+        if let Some(func_name) = FUNCTION_TO_NAME.get(self) {
+            write!(f, "{}", func_name)
+        } else {
+            // Should not be reached
+            panic!("Internal error: {self:?} not in ALL_FUNCTIONS list");
```

Comment:
Everything is great and needed, one nit, do we need to panic?

Reply:
This panic will occur if someone adds a new BuiltInFunction and doesn't add it to the built in list.

I can remove the panic but then if someone doesn't add the function to the new LIST I was worried the error would be masked (the wrong function name would be used or the lookup might fail). But maybe that is preferable 🤔

Reply:
I'm thinking wouldn't be that better to implement a function name
```
fn name(func: BuiltinScalarFunction) -> Result<String> {
 match func {
   BuiltinScalarFunction::Abs => "abs"
   ....
 }
}

This will ensure on compile time the newly added function gets string alias
```

Then we can build the hashmaps the same way over iterating BuiltinScalarFunction enum

```
    for func in BuiltinScalarFunction::iter() {
        let name = name(func);
        hashmapLookup.insert(name, func);
        hashmapRevLookup(func, name);
    }
```

Reply:
> I can play with this and create another PR

Thank you 🙏 

Note one gotcha you might hit is that the mapping is one-> many

So you might need something like 

```sql
/// returns all the names that can be used for this function
fn names(func: BuiltinScalarFunction) -> &[&'str]] {
 match func {
   BuiltinScalarFunction::Abs => "abs"
   ....
 }
}
```

Reply:
If I'm getting this correct, the only problem for now is: when adding a new function, if it's missed in `ALL_FUNCTIONS`, the compiler won't notice, so it's better to have matching arms for all `BuiltinScalarFunction` variants for `Display`.
I think this implementation works (though a bit lengthy)
```rust
impl fmt::Display for BuiltinScalarFunction {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self { // Ensure no variant is missing when adding new function
            BuiltinScalarFunction::Abs
            | BuiltinScalarFunction::Acos
            // ... add all other variants here ...
            | BuiltinScalarFunction::MakeArray => {
                if let Some(func_name) = FUNCTION_TO_NAME.get(self) {
                    write!(f, "{}", func_name)
                } else {
                    // This branch should be unreachable if the ALL_FUNCTIONS is complete
                    panic!("Internal error: {self:?} not in ALL_FUNCTIONS list", self);
                }
            },
        }
    }
}
```